### DISCUSSION TITLE
quarkus-oidc and scala extension integrations

### DISCUSSION THREAD

**Discussion by trym-moeller** [quarkusio/quarkus#37979] [Q&A] [answered]
I am trying the https://quarkus.io/guides/security-keycloak-authorization guide on my quarkus scala project. 

I get the below exception when starting the quarkus application and I would like input for how to proceed in hunting the cause for this. 

I have no knowledge of vertx. I am using the io.quarkus:quarkus-resteasy-reactive-jackson dependency and have configured the io.quarkus.jackson.ObjectMapperCustomizer with the com.fasterxml.jackson.module.scala.DefaultScalaModule, so my rest endpoints worked (before adding the oidc extension).

```
2024-01-02 15:39:51,126 WARN  [io.qua.oid.com.run.OidcCommonUtils] (vert.x-eventloop-thread-1) OIDC Server is not available:: io.vertx.core.json.DecodeException: Invalid Json Object decoded as scala.collection.immutable.HashMap
        at io.vertx.ext.web.codec.impl.BodyCodecImpl.lambda$static$1(BodyCodecImpl.java:48)
        at io.vertx.ext.web.client.HttpResponse.bodyAsJsonObject(HttpResponse.java:151)
        at io.vertx.mutiny.ext.web.client.HttpResponse.bodyAsJsonObject(HttpResponse.java:244)
        at io.quarkus.oidc.common.runtime.OidcCommonUtils.lambda$discoverMetadata$2(OidcCommonUtils.java:439)
        at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:36)
        at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:35)
        at io.smallrye.mutiny.vertx.DelegatingHandler.handle(DelegatingHandler.java:25)
        at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:397)
        at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:384)
        at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:362)
        at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:329)
        at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:291)
        at io.vertx.ext.web.client.impl.HttpContext.lambda$null$7(HttpContext.java:507)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:259)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:43)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:842)
```

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@trym-moeller AFAIK Vert.x  JsonObject is using Jackson internally and looks like the Scala mapper is interfering. My guess that having a reproducer will help, please create the one. CC @tsegismont @cescoffier

**Reply by trym-moeller** [reply]
Thanks for your suggestion. I have created the following reproducer example. https://github.com/trym-moeller/quarkus-reproducer-37979

I notice that the problem only occurs when I add a dependency to 'com.fasterxml.jackson.module:jackson-module-scala_2.13:2.16.0'
otherwise the oidc extension connects error-free to keycloak when I start the application using _./gradlew quarkusDev_. I need the dependency otherwise quarkus cannot serialize the Scala collection objects that I return in other rest endpoints in my application, I have not included an example of returning a scala collection in the reproducer.

Any ideas of how I can continue my search for the problem?

Thanks in advance for any input.

**Reply by trym-moeller** [reply]
I am considering if the "lifecycle" of the oidc extension causes problems. If the extension connects to keycloak before the json scala mapper is in place?

**Reply by sberyozkin** [reply]
@trym-moeller I don't know what can be done here. `quarkus-oidc` can't keep track of various external Jackson customizations. FYI, `quarkus-oidc` works fine with `quarkus-resteasy-reactive-jackson`, we have tests where a custom Jackson mapper is registered with `quarkus-oidc` working fine. 
Looks like `com.fasterxml.jackson.module:jackson-module-scala` has a rather global impact.

Perhaps you should checkout the Vert.x source and see why `io.vertx.core.json.DecodeException: Invalid Json Object decoded as scala.collection.immutable.HashMap` is reported and then may be tune `com.fasterxml.jackson.module:jackson-module-scala` to fix it.

I can't think of anything else that can resolve it.

**Reply by tsegismont** [reply]
@trym-moeller can you please also add instructions for configuring and running the Keycloak instance in a container please?

**Reply by sberyozkin** [reply]
@tsegismont Hi Thomas, if you'd like to try the reproducer, you can start with empty properties and `mvn quarkus:dev` which will start a container

**Reply by trym-moeller** [reply]
Hi @tsegismont

Did you succeed reproducing the behavior or can I be of further assistence?

Thanks.

**Reply by trym-moeller** [reply]
@sberyozkin The issue seems to be with vertx and jackson-module-scala as you said.

I have tried to add the BodyCodecImpl class to my project (overriding the original jar file) and it is clear that an objectmapper using the  DefaultScalaModule has already decoded the json object to a scala hashmap before the vertx BodyCodecImpl class gets to do its job. 

Why vertx ask BodyCodecImpl to do the decoding in this case and not when executing normal rest endpoints, I do not know...

Anyone with a little vertx knowledge please?

**Comment by sberyozkin** [comment]
@trym-moeller You can get the vertx source and debug and see what is going on. My guess is the Scala mapper affects the client side Jackson processing of Vertx Web Client.

**Reply by trym-moeller** [reply]
I have gotten a little more knowledge after debugging...
If I add the following code to ObjectMapperCustomizer before adding the DefaultScalaModule, then the oidc extension connects successfully to keycloak.
```Json.decodeValue("{}")```

It seems that io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec makes a copy of the quarkus managed objectmapper and adds its own (de)serializer modules, thus ensuring that the io.vertx.ext.web.codec.impl.BodyCodecImpl behaves. But when I add other deserializers (DefaultScalaModule), then the BodyCodecImpl misbehaves. Seems fragile, but what do I know.

Any suggestions for further investigations?

**Reply by sberyozkin** [reply]
@trym-moeller thanks,  looks like you have a workaround which is a good news.

**Reply by trym-moeller** [reply]
Well, not quite. Because when I then make a curl to the rest endpoint, I get the following exception
```
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class io.quarkus.security.runtime.QuarkusIdentityProviderManagerImpl$1 and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.acme.UserResource$User["identity"]->io.quarkus.security.runtime.SecurityIdentityProxy_ClientProxy["attri
```
(sorry for not having the full stack stace, the logger goes to std out, and it seems to be cut off).

**Comment by sberyozkin** [comment]
@trym-moeller So it is obvious the Scala serializer is conflicting with what Vertx `JsonObject` has to do. I wonder if you can create a much simpler reproducer, without even any security involved, for example, have a simple JAX-RS endpoint which uses Vert.x `JsonObject` to parse incoming JSON string and write JSON out, and have the Scala (de)serializer registered too, and if you can reproduce the failure, while also confirming that `quarkus` jackson modules do work as expected (example, the endpoint accepting bean `A` gets converted correctly, `quarkus-resteasy-reactive-jackson` extension supports it) then it may be a good case for a Vert.x specific bug or enhancement request

**Comment by tsegismont** [comment]
The `JSON_OBJECT_DECODER` in `io.vertx.ext.web.codec.impl.BodyCodecImpl` invokes the `Json.decodeValue`:

```java
  public static final Function<Buffer, JsonObject> JSON_OBJECT_DECODER = buff -> {
    Object val = Json.decodeValue(buff);
    if (val == null) {
      return null;
    }
    if (val instanceof JsonObject) {
      return (JsonObject) val;
    }
    throw new DecodeException("Invalid Json Object decoded as " + val.getClass().getName());
  };
```

The contract for this method is to return a `JsonObject`, `JsonArray`, `String`... etc, depending on the json content:

```java
  /**
   * Decode a given JSON buffer.
   *
   * @param buf the JSON buffer.
   *
   * @return a JSON element which can be a {@link JsonArray}, {@link JsonObject}, {@link String}, ...etc if the buffer contains an array, object, string, ...etc
   * @throws DecodeException when there is a parsing or invalid mapping.
   */
  public static Object decodeValue(Buffer buf) throws DecodeException {
    return decodeValue(buf, Object.class);
  }

  /**
   * Decode a given JSON buffer to a POJO of the given class type.
   * @param buf the JSON buffer.
   * @param clazz the class to map to.
   * @param <T> the generic type.
   * @return an instance of T
   * @throws DecodeException when there is a parsing or invalid mapping.
   */
  public static <T> T decodeValue(Buffer buf, Class<T> clazz) throws DecodeException {
    return CODEC.fromBuffer(buf, clazz);
  }
```

To do so, it invokes the `CODEC` using the `Object` class.

This works in all cases with the `io.vertx.core.json.jackson.JacksonCodec` but, with the `io.vertx.core.json.jackson.DatabindCodec` or the `io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec`, it may not work if the mapper is configured to parse something that does not implement `Map` for json objects or `List` for json arrays, like the Scala module does.

I'll try to make a reproducer test in Vert.x core and file an issue. I'll keep you informed.

**Reply by sberyozkin** [reply]
Great analysis,  thanks @tsegismont

**Reply by tsegismont** [reply]
I've created this simple Scala program:

```scala
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import io.vertx.core.json.Json
import io.vertx.core.json.jackson.DatabindCodec

object Hello {
  def main(args: Array[String]): Unit = {
    DatabindCodec.mapper().registerModule(new DefaultScalaModule)
    val value = Json.decodeValue("{\"foo\": \"baz\"}")
    println(value.getClass)
  }
}
```

It prints `class scala.collection.immutable.Map$Map1` with unmodified Vert.x core.

After applying this patch to Vert.x core:

```patch
Index: src/main/java/io/vertx/core/json/jackson/DatabindCodec.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/io/vertx/core/json/jackson/DatabindCodec.java b/src/main/java/io/vertx/core/json/jackson/DatabindCodec.java
--- a/src/main/java/io/vertx/core/json/jackson/DatabindCodec.java	(revision 74a8cc6d673b6485cc87b300f434d0ba53b5b9d0)
+++ b/src/main/java/io/vertx/core/json/jackson/DatabindCodec.java	(date 1704477174862)
@@ -16,20 +16,14 @@
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializationFeature;
-import com.fasterxml.jackson.databind.module.SimpleModule;
 import io.netty.buffer.ByteBufInputStream;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.buffer.impl.BufferInternal;
 import io.vertx.core.json.DecodeException;
 import io.vertx.core.json.EncodeException;
-import io.vertx.core.json.JsonArray;
-import io.vertx.core.json.JsonObject;
 
 import java.io.IOException;
 import java.io.InputStream;
-import java.time.Instant;
-import java.util.List;
-import java.util.Map;
 
 /**
  * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
@@ -69,21 +63,21 @@
     return prettyMapper;
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public <T> T fromValue(Object json, Class<T> clazz) {
-    T value = DatabindCodec.mapper.convertValue(json, clazz);
     if (clazz == Object.class) {
-      value = (T) adapt(value);
+      return (T) fromBuffer(toBuffer(json), Object.class);
     }
-    return value;
+    return DatabindCodec.mapper.convertValue(json, clazz);
   }
 
+  @SuppressWarnings("unchecked")
   public <T> T fromValue(Object json, TypeReference<T> type) {
-    T value = DatabindCodec.mapper.convertValue(json, type);
     if (type.getType() == Object.class) {
-      value = (T) adapt(value);
+      return (T) fromBuffer(toBuffer(json), Object.class);
     }
-    return value;
+    return DatabindCodec.mapper.convertValue(json, type);
   }
 
   @Override
@@ -120,7 +114,11 @@
     }
   }
 
+  @SuppressWarnings("unchecked")
   public static <T> T fromParser(JsonParser parser, Class<T> type) throws DecodeException {
+    if (type == Object.class) {
+      return (T) JacksonCodec.fromParser(parser, Object.class);
+    }
     T value;
     JsonToken remaining;
     try {
@@ -134,25 +132,21 @@
     if (remaining != null) {
       throw new DecodeException("Unexpected trailing token");
     }
-    if (type == Object.class) {
-      value = (T) adapt(value);
-    }
     return value;
   }
 
+  @SuppressWarnings("unchecked")
   private static <T> T fromParser(JsonParser parser, TypeReference<T> type) throws DecodeException {
-    T value;
+    if (type.getType() == Object.class) {
+      return (T) JacksonCodec.fromParser(parser, Object.class);
+    }
     try {
-      value = DatabindCodec.mapper.readValue(parser, type);
+      return DatabindCodec.mapper.readValue(parser, type);
     } catch (Exception e) {
       throw new DecodeException("Failed to decode:" + e.getMessage(), e);
     } finally {
       close(parser);
     }
-    if (type.getType() == Object.class) {
-      value = (T) adapt(value);
-    }
-    return value;
   }
 
   @Override
@@ -174,20 +168,4 @@
       throw new EncodeException("Failed to encode as JSON: " + e.getMessage());
     }
   }
-
-  private static Object adapt(Object o) {
-    try {
-      if (o instanceof List) {
-        List list = (List) o;
-        return new JsonArray(list);
-      } else if (o instanceof Map) {
-        @SuppressWarnings("unchecked")
-        Map<String, Object> map = (Map<String, Object>) o;
-        return new JsonObject(map);
-      }
-      return o;
-    } catch (Exception e) {
-      throw new DecodeException("Failed to decode: " + e.getMessage());
-    }
-  }
 }
```

The program prints `class io.vertx.core.json.JsonObject`.

This confirms the assumption in my previous comment.

Note that if Vert.x core is patched, Quarkus must be patched too (because it uses its own JsonCodec implementation).

**Reply by trym-moeller** [reply]
@tsegismont , 

This seems like the correct solution, thanks for your investigation and work! Should I create an issue in https://github.com/eclipse-vertx/vert.x/issues ? 

@sberyozkin Can I create a feature request for the io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCode to support the jackson DefaultScalaModule json (de)serializer?

**Comment by tsegismont** [comment]
I would wait before doing that. We could have a solution entirely upstream

Le lun. 8 janv. 2024, 09:50, Trym R. Møller ***@***.***> a
écrit :

**Reply by trym-moeller** [reply]
@tsegismont I am not quite sure what that means. But until I hear otherwise, I will not create an issue in vert.x. Thanks again for your input.

**Reply by tsegismont** [reply]
Sorry I made a confusion, I initially thought you wanted to open an issue for tracking the JsonObject decoding problem on Quarkus.

**Reply by trym-moeller** [reply]
That is quite allright. Then I believe that you suggest to wait for a fix in vert.x that quarkus can utilize when upgrading its dependency to vert.x (and fix the QuarkusJacksonJsonCodec at update time as well).

Do you need me to create an issue in the vert.x github issue tracker?

**Reply by trym-moeller** [reply]
Hi Sergey @sberyozkin 

I am unsure how to proceed from here. Should I create an issue regarding the QuarkusJacksonJsonCodec class or?
Thanks for you input.

Best regards Trym

**Reply by tsegismont** [reply]
Please don't. Let me finish figuring out what we can do upstream.

**Reply by trym-moeller** [reply]
Ok, good to know, that you are working on a solution.

**Reply by tsegismont** [reply]
Filed https://github.com/eclipse-vertx/vert.x/issues/5058 and sent PR https://github.com/eclipse-vertx/vert.x/pull/5057

When we come to an agreement on the solution upstream, I will file an issue in Quarkus and send a PR

**Comment by tsegismont** [comment]
@trym-moeller can you please try using this module instead of the default Scala module:

```scala
import com.fasterxml.jackson.module.scala._
import com.fasterxml.jackson.module.scala.deser._
import com.fasterxml.jackson.module.scala.introspect._

class MyScalaModule
  extends JacksonModule
    with IteratorModule
    with EnumerationModule
    with OptionModule
    with SeqModule
    with IterableModule
    with TupleModule
    with MapModule
    with SetModule
    with ScalaNumberDeserializersModule
    with ScalaObjectDeserializerModule
    with ScalaAnnotationIntrospectorModule
    with EitherModule
    with SymbolModule {
  override def getModuleName: String = "MyScalaModule"
}

object MyScalaModule
```

The difference is it doesn't add `com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializerModule`

**Reply by trym-moeller** [reply]
Hi Thomas @tsegismont 

I have tried your suggestion (added it to the reproducer in https://github.com/trym-moeller/quarkus-reproducer-37979), and it fails as originally with the _io.vertx.core.json.DecodeException: Invalid Json Object decoded as scala.collection.immutable.HashMap_ exception.

Thanks for your try!

Best regards Trym

**Reply by tsegismont** [reply]
I gave it a try again, this is because the Quarkus Scala extension auto loads `DefaultScalaModule`

https://github.com/quarkusio/quarkus/blob/235bff398e8b1ebc741d9e09633262a4a313c822/extensions/scala/deployment/src/main/java/io/quarkus/scala/deployment/ScalaProcessor.java#L10-L32

**Reply by trym-moeller** [reply]
Hi Thomas @tsegismont ,

Good catch!
I have updated the reproducer with an override of the io.quarkus.scala.deployment.ScalaProcessor class (not adding the DefaultScalaModule), so the proposed modified jackson default scala module is added instead. Now the first oidc "handshake" works out, but when I try to call the rest endpoint, the communication fails with:
`2024-01-11 09:34:21,919 ERROR [io.qua.res.rea.jac.run.map.NativeInvalidDefinitionExceptionMapper] (executor-thread-1) com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class io.quarkus.security.runtime.QuarkusIdentityProviderManagerImpl$1 and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.acme.UserResource$User["identity"]->io.quarkus.security.runtime.SecurityIdentityProxy_ClientProxy["attributes"]->java.util.Collections$UnmodifiableMap["io.quarkus.security.identity.AuthenticationRequestContext"])`

I am unsure what this indicates?

Best regards Trym

**Reply by tsegismont** [reply]
Perhaps your custom Scala processor is missing this piece:

```project
@BuildStep
    FeatureBuildItem feature() {
        return new FeatureBuildItem(Feature.SCALA);
    }
```

**Reply by trym-moeller** [reply]
@tsegismont, sorry for not reading my last message through.

I tried to json serialize the QuarkusIdentityProviderManagerImpl class (and there were no json serializer for that class). I have now modified my reproducer, so it does not serialize this class... and everything (until now) is working as expected.

I am unsure what consequences to two hacks have, but I will try to include them in my original project and see.

Thanks for you help until now. I am unsure, if the hacks should belong to me alone or if we can contribute them to the quarkus code base?

Best regards Trym

**Comment by tsegismont** [comment, marked as answer]
> I am unsure what consequences to two hacks have, but I will try to include them in my original project and see.
> 
> Thanks for you help until now. I am unsure, if the hacks should belong to me alone or if we can contribute them to the quarkus code base?

I don't know about the consequences of overriding a processor. Ideally, a contribution to the Scala extension would be better. But then the Scala extension you are using is deprecated, so I don't believe an enhancement PR would be accepted.

Maybe @gsmet has some thoughts about this?

On using a custom Jackson Scala module, the `DefaultScalaModule` doc reads:

"_This class aggregates all of the feature modules into a single concrete class. Its use is recommended for new users and users who want things to "just work". If more customized support is desired, consult each of the constituent traits._"

The `com.fasterxml.jackson.module.scala.deser.UntypedObjectDeserializerModule` that I suggested to remove is used when Jackson Databind has no idea of the target type for deserialization. When this happens, it will use the default, which is to deserialize to Java Map or List for Json objects or arrays.

So I believe you should be fine if you deserialize JSON payloads to your own Scala types (like your `User` type).

**Comment by sberyozkin** [comment]
@tsegismont, major thanks for working with @trym-moeller and finding a solution,  apologies I was not replying, I was on PTO

**Reply by tsegismont** [reply]
You're welcome
